File: src/commands/moderation/unban.ts
import { SlashCommandBuilder, PermissionFlagsBits, ChatInputCommandInteraction } from 'discord.js';

export default {
  data: new SlashCommandBuilder()
    .setName('unban')
    .setDescription('Unban a user from the server')
    .setDefaultMemberPermissions(PermissionFlagsBits.BanMembers)
    .addStringOption(option =>
      option.setName('userid').setDescription('User ID to unban').setRequired(true)
    )
    .addStringOption(option =>
      option.setName('reason').setDescription('Reason for unban').setRequired(false)
    ),
  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    if (!interaction.guild) return;
    const userId = interaction.options.getString('userid', true);
    const reason = interaction.options.getString('reason') || 'No reason provided';

    if (!interaction.guild.members.me?.permissions.has(PermissionFlagsBits.BanMembers)) {
      await interaction.reply({ content: 'I need ban permissions to do that.', ephemeral: true });
      return;
    }

    try {
      const bans = await interaction.guild.bans.fetch();
      const bannedUser = bans.get(userId);
      if (!bannedUser) {
        await interaction.reply({ content: 'That user is not banned.', ephemeral: true });
        return;
      }

      await interaction.guild.bans.remove(userId, reason);
      await interaction.reply({
        content: `Unbanned user ${bannedUser.user.tag} | Reason: ${reason}`,
        ephemeral: true
      });
    } catch (err) {
      await interaction.reply({
        content: `Failed to unban user: ${(err as Error).message}`,
        ephemeral: true
      });
    }
  }
};

File: src/bot/commands/moderation/unban.rs
use crate::bot::{Context, Error};
use poise::serenity_prelude as serenity;

/// Unban a user from the server
#[poise::command(
    slash_command,
    ephemeral,
    guild_only,
    required_permissions = "BAN_MEMBERS",
    required_bot_permissions = "BAN_MEMBERS"
)]
pub async fn unban(
    ctx: Context<'_>,
    #[description = "Discord User ID to unban"] user_id: String,
    #[description = "Reason for unban"] reason: Option<String>,
) -> Result<(), Error> {
    let reason = reason.as_deref().unwrap_or("No reason provided");
    // Extract guild_id (Copy) before any await
    let guild_id = ctx
        .guild_id()
        .ok_or_else(|| anyhow::anyhow!("Not in a guild"))?;

    let uid: u64 = user_id
        .trim()
        .parse()
        .map_err(|_| anyhow::anyhow!("Invalid user ID – must be a numeric Discord snowflake"))?;

    // Snowflake 0 is never a real user
    if uid == 0 {
        ctx.say("Invalid user ID.").await?;
        return Ok(());
    }

    let uid = serenity::UserId::new(uid);

    // Look up the specific ban entry directly instead of fetching all bans
    // (fetching all bans silently fails on servers with > 1000 bans).
    let ban = match guild_id.ban(ctx, uid).await {
        Ok(b) => b,
        Err(_) => {
            ctx.say("That user is not banned on this server.").await?;
            return Ok(());
        }
    };

    guild_id
        .unban(ctx, uid)
        .await
        .map_err(|e| anyhow::anyhow!("Failed to unban: {e}"))?;

    ctx.say(format!("Unbanned {} | Reason: {reason}", ban.user.name))
        .await?;
    Ok(())
}
